File: java/var-core/src/test/java/dev/varar/core/RegistryTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.cucumber.cucumberexpressions.Argument;
import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/registry.test.ts}. */
class RegistryTest {

    private static final Object NOOP_HANDLER = (Runnable) () -> {};

    @Test
    void createRegistryReturnsAnEmptyRegistryWithDefaultParameterTypes() {
        Registry r = Registry.createRegistry();
        assertEquals(0, r.steps().size());
        assertNotNull(r.parameterTypes());
    }

    @Test
    void addStepReturnsANewRegistryOriginalIsUnchanged() {
        Registry r0 = Registry.createRegistry();
        Registry r1 = Registry.addStep(r0, "I have {int} cukes", "steps.ts", 1, NOOP_HANDLER, null);

        assertEquals(0, r0.steps().size());
        assertEquals(1, r1.steps().size());
        assertEquals("I have {int} cukes", r1.steps().get(0).expression());
    }

    @Test
    void defineParameterTypeMakesACustomTypeAvailableToSubsequentStepCompilations() {
        Registry r = Registry.createRegistry();
        Registry withType =
                Registry.defineParameterType(r, "airport", Pattern.compile("[A-Z]{3}"), groups -> groups[0]);

        // Compiling an expression that uses {airport} should now succeed without an
        // UndefinedParameterTypeError.
        assertDoesNotThrow(() -> Registry.addStep(withType, "I fly to {airport}", "steps.ts", 1, NOOP_HANDLER, null));
    }

    @Test
    void defineParameterTypeReturnedStepActuallyMatchesTheRegexAtRuntime() {
        Registry r = Registry.createRegistry();
        r = Registry.defineParameterType(r, "airport", Pattern.compile("[A-Z]{3}"), groups -> groups[0].toLowerCase());
        r = Registry.addStep(r, "I fly to {airport}", "steps.ts", 1, NOOP_HANDLER, null);

        Optional<List<Argument<?>>> match = r.steps().get(0).compiled().match("I fly to LHR");
        assertTrue(match.isPresent());
        assertEquals("lhr", match.get().get(0).getValue());
    }

    @Test
    void addStepThrowsOnDuplicateExpressionsListingBothSourcePositions() {
        Registry r = Registry.createRegistry();
        Registry withFirst = Registry.addStep(r, "I have {int} cukes", "a.ts", 3, NOOP_HANDLER, null);

        IllegalArgumentException ex = assertThrows(
                IllegalArgumentException.class,
                () -> Registry.addStep(withFirst, "I have {int} cukes", "b.ts", 9, NOOP_HANDLER, null));
        assertTrue(ex.getMessage().matches("(?s).*duplicate step definition.*a\\.ts:3.*b\\.ts:9.*"));
    }

    @Test
    void addStepCarriesTheStepKindThroughToTheRegistration() {
        Registry r = Registry.addStep(
                Registry.createRegistry(), "I greet {string}", "a.steps.ts", 1, NOOP_HANDLER, StepKind.SENSOR);
        assertEquals(StepKind.SENSOR, r.steps().get(0).kind());
    }

    @Test
    void kindIsOptionalLegacyStepPath() {
        Registry r =
                Registry.addStep(Registry.createRegistry(), "I greet {string}", "a.steps.ts", 1, NOOP_HANDLER, null);
        assertNull(r.steps().get(0).kind());
    }

    @Test
    void defineParameterTypeRecordsTheCustomTypeImmutably() {
        Registry r0 = Registry.createRegistry();
        assertEquals(List.of(), r0.customParameterTypes());
        Registry r1 =
                Registry.defineParameterType(r0, "airport", java.util.regex.Pattern.compile("[A-Z]{3}"), g -> g[0]);
        assertEquals(List.of(new Registry.CustomParameterType("airport", "[A-Z]{3}")), r1.customParameterTypes());
        // The original registry value is untouched (records are immutable views).
        assertEquals(List.of(), r0.customParameterTypes());
        assertThrows(
                UnsupportedOperationException.class,
                () -> r1.customParameterTypes().add(new Registry.CustomParameterType("x", "y")));
    }
}

File: rust/core/tests/registry_test.rs
//! Port of `RegistryTest.java` / `registry.test.ts`. Java's `Pattern.compile(...)`
//! becomes a bare regexp `&str`; the `assertThrows(UnsupportedOperationException)`
//! immutability clause is dropped (Rust values are immutable).

use std::rc::Rc;
use varar_core::error::RegistryError;
use varar_core::handler::Handler;
use varar_core::registry::{CustomParameterType, add_step, create_registry, define_parameter_type};
use varar_core::step_kind::StepKind;
use varar_core::value::Value;

#[test]
fn create_registry_returns_an_empty_registry_with_default_parameter_types() {
    let r = create_registry();
    assert_eq!(0, r.steps.len());
}

#[test]
fn add_step_returns_a_new_registry_original_is_unchanged() {
    let r0 = create_registry();
    let r1 = add_step(&r0, "I have {int} cukes", "steps.ts", 1, Handler::noop(), None).unwrap();
    assert_eq!(0, r0.steps.len());
    assert_eq!(1, r1.steps.len());
    assert_eq!("I have {int} cukes", r1.steps[0].expression);
}

#[test]
fn define_parameter_type_makes_a_custom_type_available_to_subsequent_step_compilations() {
    let r = create_registry();
    let with_type =
        define_parameter_type(&r, "airport", "[A-Z]{3}", Rc::new(|g: &[&str]| Value::from(g[0])));
    assert!(
        add_step(&with_type, "I fly to {airport}", "steps.ts", 1, Handler::noop(), None).is_ok()
    );
}

#[test]
fn define_parameter_type_returned_step_actually_matches_the_regex_at_runtime() {
    let r = create_registry();
    let r = define_parameter_type(
        &r,
        "airport",
        "[A-Z]{3}",
        Rc::new(|g: &[&str]| Value::from(g[0].to_lowercase())),
    );
    let r = add_step(&r, "I fly to {airport}", "steps.ts", 1, Handler::noop(), None).unwrap();
    let matched = r.steps[0].compiled.match_whole("I fly to LHR");
    assert!(matched.is_some());
    assert_eq!(Value::from("lhr"), matched.unwrap()[0].value);
}

#[test]
fn add_step_throws_on_duplicate_expressions_listing_both_source_positions() {
    let r = create_registry();
    let with_first = add_step(&r, "I have {int} cukes", "a.ts", 3, Handler::noop(), None).unwrap();
    let err = match add_step(&with_first, "I have {int} cukes", "b.ts", 9, Handler::noop(), None) {
        Ok(_) => panic!("expected a duplicate-step error"),
        Err(e) => e,
    };
    let RegistryError::DuplicateStep(msg) = err else {
        panic!("expected duplicate error")
    };
    assert!(msg.contains("duplicate step definition"));
    assert!(msg.contains("a.ts:3"));
    assert!(msg.contains("b.ts:9"));
}

#[test]
fn add_step_carries_the_step_kind_through_to_the_registration() {
    let r = add_step(
        &create_registry(),
        "I greet {string}",
        "a.steps.ts",
        1,
        Handler::noop(),
        Some(StepKind::Sensor),
    )
    .unwrap();
    assert_eq!(Some(StepKind::Sensor), r.steps[0].kind);
}

#[test]
fn kind_is_optional_legacy_step_path() {
    let r =
        add_step(&create_registry(), "I greet {string}", "a.steps.ts", 1, Handler::noop(), None)
            .unwrap();
    assert_eq!(None, r.steps[0].kind);
}

#[test]
fn define_parameter_type_records_the_custom_type_immutably() {
    let r0 = create_registry();
    assert_eq!(Vec::<CustomParameterType>::new(), r0.custom_parameter_types);
    let r1 =
        define_parameter_type(&r0, "airport", "[A-Z]{3}", Rc::new(|g: &[&str]| Value::from(g[0])));
    assert_eq!(vec![CustomParameterType::new("airport", "[A-Z]{3}")], r1.custom_parameter_types);
    // The original registry value is untouched.
    assert_eq!(Vec::<CustomParameterType>::new(), r0.custom_parameter_types);
}
